SWIG Interfaces, Python bindings and tests for the Python bindings

MichaelClerx opened the pull request:
This PR:

- Adds infrastructure for SWIG bindings
- Adds general purpose (multi-language) SWIG interface files
- Uses those interface files to generate Python bindings (#161)
- Creates tests (integrated with the ctest framework) of the Python bindings, which serve as tests for the SWIG bindings in general
- Adds docstrings to the bindings (#208)

MichaelClerx committed:
Added infra for Python bindings

hsorby committed:
An initial stab at adding Python bindings for libcellml

This initial attempt is by no means complete but I think it is heading in the right direction.  Things to improve upon would be typemaps for methods returning values like units.getUnitAttributes method.

hsorby committed:
Remove commented out code.

MichaelClerx committed:
Added infra for python binding tests

MichaelClerx committed:
Added tests for entity and named entity

MichaelClerx committed:
Working on tests basic entity classes

MichaelClerx committed:
Added some support for shared_ptrs

MichaelClerx committed:
wip

MichaelClerx committed:
Finished test_imported_entity

MichaelClerx committed:
wip

MichaelClerx committed:
Added tests for inheritance

MichaelClerx committed:
Fixed issue with Entity ptr following hints from @hsorby and @agarny

MichaelClerx committed:
Changed cmake option to LIBCELLML_BINDINGS_PYTHON

MichaelClerx committed:
Marked SWIG_EXECUTABLE option as 'advanced' in CMake

MichaelClerx committed:
Python binding tests are now written to build/tests

MichaelClerx committed:
Moved Python binding tests to tests directory

hsorby committed:
Add infrastructure to allow tests to be run individually.  Rename test classes to use *TestCase.

hsorby committed:
Put the LIBCELLML_BINDINGS_PYTHON into the top level.

hsorby committed:
Remove test_bindings_python.py script.

hsorby committed:
Add properties to tests for Windows.

hsorby committed:
Move add_subdirectory command for bindings into src directory.

hsorby committed:
Explicitly ignore methods for SWIG.

Explicitly tell SWIG to ignore the move constructor and assignment operator.

MichaelClerx committed:
Fixed issue where binding tests weren't actually run

hsorby committed:
Fix merge conflicts.

hsorby committed:
Look for LibXml2 queitly in config mode on Windows.

hsorby committed:
Add missing 'import sys' statement to Python tests.

hsorby committed:
Re-instate suite method to Python tests.

hsorby committed:
Fix indentation in tests/bindings/python/CMakeLists.txt.

hsorby committed:
Revert to using unittest.main().

MichaelClerx committed:
Added tests for ImportSource

MichaelClerx committed:
More tests for ComponentEntity bindings.

MichaelClerx committed:
Finished tests for ComponentEntity

MichaelClerx committed:
Add swig Variable interface

MichaelClerx committed:
Structured test_entity.py

MichaelClerx committed:
Structured test_named_entity.py

MichaelClerx committed:
Structured test_imported_entity.py

MichaelClerx committed:
Structured test_import_source.py

MichaelClerx committed:
Finished test_component.py

MichaelClerx committed:
Python wrappers: Added tests and handling for cases where methods x(size_t) were called with negative ints.

MichaelClerx committed:
Added interface and tests for model

MichaelClerx committed:
Added initial test_units.py

MichaelClerx committed:
Finished tests for Variable bindings

MichaelClerx committed:
Added classes to main libcellml module.

MichaelClerx committed:
Bindings: Fixed naming of converted enums

MichaelClerx committed:
Bindings: Working on Units bindings

MichaelClerx committed:
Bindings: Finished interface & tests for units

MichaelClerx committed:
[bindings] Adding framework for more classes

MichaelClerx committed:
[bindings] Added more binding files (not complete yet)

MichaelClerx committed:
[bindings] Added more enumerations

MichaelClerx committed:
[bindings] Added tests for Error bindings

MichaelClerx committed:
[bindings] Minor fixes and cleanup

MichaelClerx committed:
[bindings] Added tests and interface for logger

MichaelClerx committed:
[bindings] Added printer interface and tests

MichaelClerx committed:
[bindings] Added parser interface and tests

MichaelClerx committed:
[bindings] Added validator interface and tests

MichaelClerx committed:
[bindings] Added version function interface and tests.

MichaelClerx committed:
Renamed getSourceImport() to getImportSource()

MichaelClerx committed:
Brought bindings in line with recent changes

MichaelClerx committed:
Fixed warnings from ImportSource bindings

hsorby committed:
Add return value on extended method in units.i.

MichaelClerx committed:
[bindings] Added print of swig version

MichaelClerx committed:
Made SWIG 3 a requirement when building the bindings

MichaelClerx committed:
[bindings] Fix for warnings on windows

MichaelClerx committed:
Windows test

MichaelClerx committed:
[bindings] Added docstring

MichaelClerx committed:
[bindings] Fixed issues with conversion of enumerations causing bug on reload(libcellml)

MichaelClerx committed:
Trying to debug ubuntu 14.04 build fails

MichaelClerx committed:
Enumerations fix for swig 3 old than 3.0.3

MichaelClerx committed:
Removed XML enum from bindings

MichaelClerx committed:
[bindings] Added docstrings to ImportedEntity, ComponentEntity, and Component

MichaelClerx committed:
[bindings] Added docstrings to Entity and ImportSource

MichaelClerx committed:
[bindings] Added docstrings to ImportedEntity and NamedEntity

MichaelClerx committed:
[bindings] Added more docstrings

MichaelClerx committed:
[bindings] Added remaining docstrings. Closes #208.

agarny commented:
Besides issue #208, this PR addresses what other issues, if any?

agarny commented:
Apparently, it also addresses issue #161? It would be nice to have all the issues this PR addresses in the top comment.

MichaelClerx commented:
Thanks @agarny, have updated the top comment!

hsorby committed:
Add typemap for checking the values of the enum.

The current ability to throw any int into a method that is expecting an enumeration is a problem for C++.  It seems the standard currently defines casting to a value outside the value of an enumeration as resulting in a value that is unspecified.  Not very helpful, but in the future this will be change and will result in undefined behaviour.  So letting this go unchecked is not going to be an option.

MichaelClerx committed:
Added typemaps for all enumerations. Hid addUnit(StandardUnit, double exponent) to avoid confusion with addUnit(StandardUnit, Prefix (int), double=1, double=1)

MichaelClerx committed:
Improved comment

nickerso commented:
@hsorby - do we know why the tests are not running on the build machines? They seem to build and run successfully on my mac.

agarny commented:
FWIW, I just had a quick look at testing this PR on my mac (10.13.2 with CMake 3.10.0) and I am several of the following message:

```
 CMake Deprecation Warning at /usr/local/Cellar/cmake/3.10.0/share/cmake/Modules/UseSWIG.cmake:231 (message):
   SWIG_ADD_MODULE is deprecated.  Use SWIG_ADD_LIBRARY instead.
 Call Stack (most recent call first):
   src/bindings/python/CMakeLists.txt:60 (swig_add_module)
```

Still, I am able to build things just fine. However, when I test them (using `make test`), I am getting the following error:

```
The following tests FAILED:
	  6 - io_unit_test_parser (Failed)
Errors while running CTest
make: *** [test] Error 8
```

P.S.: just in case, I did a simple `git clone https://github.com/MichaelClerx/libcellml.git` and then `git checkout issue161`.

MichaelClerx commented:
Thanks @agarny ! I'm getting the deprecation warnings now too (although they weren't there when we worked on this in October), will look into those today!

For the error, could you run:
```
env CTEST_OUTPUT_ON_FAILURE=1 ctest
```
to see the error output?

Incidentally, does anyone know if there's a way to configure our project so that ctest always does this? Very strange design decision to hide the error output of a testing program :D :D: D

agarny commented:
Just out of curiosity, in commit 753f040, why do you not simply use `swig_add_library()`, no matter the version of CMake?

Anyway, here is what I got for the tests:

```
Alan@Venus:~/Desktop/libcellml/build (issue161)$ env CTEST_OUTPUT_ON_FAILURE=1 ctest
Test project /Users/Alan/Desktop/libcellml/build
      Start  1: entities_unit_test_component
 1/26 Test  #1: entities_unit_test_component .....   Passed    0.02 sec
      Start  2: entities_unit_test_connection
 2/26 Test  #2: entities_unit_test_connection ....   Passed    0.02 sec
      Start  3: misc_unit_test_coverage
 3/26 Test  #3: misc_unit_test_coverage ..........   Passed    0.02 sec
      Start  4: entities_unit_test_math
 4/26 Test  #4: entities_unit_test_math ..........   Passed    0.02 sec
      Start  5: entities_unit_test_model
 5/26 Test  #5: entities_unit_test_model .........   Passed    0.02 sec
      Start  6: io_unit_test_parser
 6/26 Test  #6: io_unit_test_parser ..............***Failed    0.05 sec
Running main() from gtest_main.cc
[==========] Running 61 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 61 tests from Parser
[ RUN      ] Parser.invalidXMLElements
/Users/Alan/Desktop/libcellml/tests/parser/parser.cpp:50: Failure
Value of: p.getError(i)->getDescription()
  Actual: "Specification mandates value for attribute bearded."
Expected: expectedErrors.at(i)
Which is: "Specification mandate value for attribute bearded."
[  FAILED  ] Parser.invalidXMLElements (0 ms)
[ RUN      ] Parser.parse
[       OK ] Parser.parse (0 ms)
[ RUN      ] Parser.parseNamedModel
[       OK ] Parser.parseNamedModel (0 ms)
[ RUN      ] Parser.moveParser
[       OK ] Parser.moveParser (0 ms)
[ RUN      ] Parser.makeError
[       OK ] Parser.makeError (0 ms)
[ RUN      ] Parser.emptyModelString
[       OK ] Parser.emptyModelString (0 ms)
[ RUN      ] Parser.nonXmlString
[       OK ] Parser.nonXmlString (0 ms)
[ RUN      ] Parser.invalidRootNode
[       OK ] Parser.invalidRootNode (0 ms)
[ RUN      ] Parser.invalidModelAttribute
[       OK ] Parser.invalidModelAttribute (0 ms)
[ RUN      ] Parser.invalidModelElement
[       OK ] Parser.invalidModelElement (0 ms)
[ RUN      ] Parser.modelWithInvalidElement
[       OK ] Parser.modelWithInvalidElement (0 ms)
[ RUN      ] Parser.parseModelWithInvalidAttributeAndGetError
[       OK ] Parser.parseModelWithInvalidAttributeAndGetError (0 ms)
[ RUN      ] Parser.parseNamedModelWithNamedComponent
[       OK ] Parser.parseNamedModelWithNamedComponent (0 ms)
[ RUN      ] Parser.parseModelWithUnitsAndNamedComponent
[       OK ] Parser.parseModelWithUnitsAndNamedComponent (0 ms)
[ RUN      ] Parser.unitsAttributeError
[       OK ] Parser.unitsAttributeError (0 ms)
[ RUN      ] Parser.unitsElementErrors
[       OK ] Parser.unitsElementErrors (1 ms)
[ RUN      ] Parser.parseModelWithNamedComponentWithInvalidBaseUnitsAttributeAndGetError
[       OK ] Parser.parseModelWithNamedComponentWithInvalidBaseUnitsAttributeAndGetError (0 ms)
[ RUN      ] Parser.parseModelWithInvalidComponentAttributeAndGetError
[       OK ] Parser.parseModelWithInvalidComponentAttributeAndGetError (0 ms)
[ RUN      ] Parser.componentAttributeErrors
[       OK ] Parser.componentAttributeErrors (0 ms)
[ RUN      ] Parser.componentElementErrors
[       OK ] Parser.componentElementErrors (0 ms)
[ RUN      ] Parser.parseModelWithTwoComponents
[       OK ] Parser.parseModelWithTwoComponents (0 ms)
[ RUN      ] Parser.parseModelWithComponentHierarchyWaterfall
[       OK ] Parser.parseModelWithComponentHierarchyWaterfall (0 ms)
[ RUN      ] Parser.parseModelWithMultipleComponentHierarchyWaterfalls
[       OK ] Parser.parseModelWithMultipleComponentHierarchyWaterfalls (0 ms)
[ RUN      ] Parser.modelWithUnits
[       OK ] Parser.modelWithUnits (0 ms)
[ RUN      ] Parser.modelWithInvalidUnits
[       OK ] Parser.modelWithInvalidUnits (0 ms)
[ RUN      ] Parser.emptyEncapsulation
[       OK ] Parser.emptyEncapsulation (0 ms)
[ RUN      ] Parser.encapsulationWithNoComponentAttribute
[       OK ] Parser.encapsulationWithNoComponentAttribute (0 ms)
[ RUN      ] Parser.encapsulationWithNoComponentRef
[       OK ] Parser.encapsulationWithNoComponentRef (0 ms)
[ RUN      ] Parser.encapsulationWithNoComponent
[       OK ] Parser.encapsulationWithNoComponent (0 ms)
[ RUN      ] Parser.encapsulationWithMissingComponent
[       OK ] Parser.encapsulationWithMissingComponent (0 ms)
[ RUN      ] Parser.encapsulationWithNoComponentChild
[       OK ] Parser.encapsulationWithNoComponentChild (0 ms)
[ RUN      ] Parser.encapsulationNoChildComponentRef
[       OK ] Parser.encapsulationNoChildComponentRef (0 ms)
[ RUN      ] Parser.encapsulationWithNoGrandchildComponentRef
[       OK ] Parser.encapsulationWithNoGrandchildComponentRef (0 ms)
[ RUN      ] Parser.invalidEncapsulations
[       OK ] Parser.invalidEncapsulations (0 ms)
[ RUN      ] Parser.invalidVariableAttributesAndGetVariableError
[       OK ] Parser.invalidVariableAttributesAndGetVariableError (1 ms)
[ RUN      ] Parser.variableAttributeAndChildErrors
[       OK ] Parser.variableAttributeAndChildErrors (0 ms)
[ RUN      ] Parser.emptyConnections
[       OK ] Parser.emptyConnections (0 ms)
[ RUN      ] Parser.connectionErrorNoComponent2
[       OK ] Parser.connectionErrorNoComponent2 (0 ms)
[ RUN      ] Parser.connectionErrorNoComponent2InModel
[       OK ] Parser.connectionErrorNoComponent2InModel (0 ms)
[ RUN      ] Parser.connectionErrorNoComponent1
[       OK ] Parser.connectionErrorNoComponent1 (0 ms)
[ RUN      ] Parser.connectionErrorNoMapComponents
[       OK ] Parser.connectionErrorNoMapComponents (0 ms)
[ RUN      ] Parser.connectionErrorNoMapVariables
[       OK ] Parser.connectionErrorNoMapVariables (0 ms)
[ RUN      ] Parser.importedComponent2Connection
[       OK ] Parser.importedComponent2Connection (0 ms)
[ RUN      ] Parser.validConnectionMapVariablesFirst
[       OK ] Parser.validConnectionMapVariablesFirst (1 ms)
[ RUN      ] Parser.component2ConnectionVariableMissing
[       OK ] Parser.component2ConnectionVariableMissing (0 ms)
[ RUN      ] Parser.component2InConnectionMissing
[       OK ] Parser.component2InConnectionMissing (0 ms)
[ RUN      ] Parser.connectionVariable2Missing
[       OK ] Parser.connectionVariable2Missing (0 ms)
[ RUN      ] Parser.connectionVariable1Missing
[       OK ] Parser.connectionVariable1Missing (0 ms)
[ RUN      ] Parser.connectionErrorNoMapVariablesType
[       OK ] Parser.connectionErrorNoMapVariablesType (0 ms)
[ RUN      ] Parser.invalidImportsAndGetError
[       OK ] Parser.invalidImportsAndGetError (0 ms)
[ RUN      ] Parser.invalidModelWithAllKindsOfErrors
[       OK ] Parser.invalidModelWithAllKindsOfErrors (0 ms)
[ RUN      ] Parser.invalidModelWithTextInAllElements
[       OK ] Parser.invalidModelWithTextInAllElements (1 ms)
[ RUN      ] Parser.parseIds
[       OK ] Parser.parseIds (0 ms)
[ RUN      ] Parser.parseSineModelFromFile
[       OK ] Parser.parseSineModelFromFile (1 ms)
[ RUN      ] Parser.parseSineImportsModelFromFile
[       OK ] Parser.parseSineImportsModelFromFile (0 ms)
[ RUN      ] Parser.parseInvalidModelFromFile
[       OK ] Parser.parseInvalidModelFromFile (1 ms)
[ RUN      ] Parser.parseOrdModelFromFile
[       OK ] Parser.parseOrdModelFromFile (22 ms)
[ RUN      ] Parser.parseComplexEncapsulationModelFromFile
[       OK ] Parser.parseComplexEncapsulationModelFromFile (0 ms)
[ RUN      ] Parser.parseModelWithComponentsWithMultipleMathElements
[       OK ] Parser.parseModelWithComponentsWithMultipleMathElements (1 ms)
[ RUN      ] Parser.parseValidXmlDirectlyUsingLibxml
[       OK ] Parser.parseValidXmlDirectlyUsingLibxml (0 ms)
[ RUN      ] Parser.parseInvalidXmlDirectlyUsingLibxml
/:2: parser error : Opening and ending tag mismatch: component line 2 and model
<model xmlns="http://www.cellml.org/cellml/2.0#"><component></model>
                                                                    ^
/:2: parser error : Premature end of data in tag model line 2
<model xmlns="http://www.cellml.org/cellml/2.0#"><component></model>
                                                                    ^
[       OK ] Parser.parseInvalidXmlDirectlyUsingLibxml (0 ms)
[----------] 61 tests from Parser (31 ms total)

[----------] Global test environment tear-down
[==========] 61 tests from 1 test case ran. (31 ms total)
[  PASSED  ] 60 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] Parser.invalidXMLElements

 1 FAILED TEST

      Start  7: io_unit_test_printer
 7/26 Test  #7: io_unit_test_printer .............   Passed    0.02 sec
      Start  8: entities_unit_test_units
 8/26 Test  #8: entities_unit_test_units .........   Passed    0.02 sec
      Start  9: io_unit_test_validator
 9/26 Test  #9: io_unit_test_validator ...........   Passed    0.10 sec
      Start 10: entities_unit_test_variable
10/26 Test #10: entities_unit_test_variable ......   Passed    0.02 sec
      Start 11: api_unit_test_version
11/26 Test #11: api_unit_test_version ............   Passed    0.02 sec
      Start 12: python_test_component
12/26 Test #12: python_test_component ............   Passed    0.14 sec
      Start 13: python_test_component_entity
13/26 Test #13: python_test_component_entity .....   Passed    0.09 sec
      Start 14: python_test_entity
14/26 Test #14: python_test_entity ...............   Passed    0.08 sec
      Start 15: python_test_error
15/26 Test #15: python_test_error ................   Passed    0.08 sec
      Start 16: python_test_import_source
16/26 Test #16: python_test_import_source ........   Passed    0.08 sec
      Start 17: python_test_imported_entity
17/26 Test #17: python_test_imported_entity ......   Passed    0.08 sec
      Start 18: python_test_logger
18/26 Test #18: python_test_logger ...............   Passed    0.08 sec
      Start 19: python_test_model
19/26 Test #19: python_test_model ................   Passed    0.09 sec
      Start 20: python_test_named_entity
20/26 Test #20: python_test_named_entity .........   Passed    0.09 sec
      Start 21: python_test_parser
21/26 Test #21: python_test_parser ...............   Passed    0.08 sec
      Start 22: python_test_printer
22/26 Test #22: python_test_printer ..............   Passed    0.08 sec
      Start 23: python_test_units
23/26 Test #23: python_test_units ................   Passed    0.09 sec
      Start 24: python_test_variable
24/26 Test #24: python_test_variable .............   Passed    0.09 sec
      Start 25: python_test_validator
25/26 Test #25: python_test_validator ............   Passed    0.08 sec
      Start 26: python_test_version
26/26 Test #26: python_test_version ..............   Passed    0.08 sec

96% tests passed, 1 tests failed out of 26

Total Test time (real) =   1.68 sec

The following tests FAILED:
	  6 - io_unit_test_parser (Failed)
Errors while running CTest
Alan@Venus:~/Desktop/libcellml/build (issue161)$ 
```

MichaelClerx commented:
Thanks! That's not a Python bug it seems, should happen on master now too!

nickerso commented:
The tests all pass for me - but rather than a fresh clone of @MichaelClerx's issue161 branch, I am merging that branch into an up-to-date `prime/develop` repo...might be worth making sure that the issue161 branch is updated from `prime/develop`?

nickerso commented:
Or I've screwed something else up - I have been having loads of "fun" with git today :)

MichaelClerx commented:
> Just out of curiosity, in commit 753f040, why do you not simply use swig_add_library(), no matter the version of CMake?
Versions before 3.8 didn't provide `swig_add_library`, so you need to have the switch.

It seems our old code always used `swig_add_module` (the check didn't work) which is why we were seeing deprecation warnings on newer systems.

MichaelClerx commented:
Could you all git pull and check the latest commit for this branch is `042f5411c11be7dd144414f06982c62b054e3253` ?

MichaelClerx commented:
@nickerso I don't really understand why you're merging instead of just checking it out? --> Ok get it now!

MichaelClerx commented:
> might be worth making sure that the issue161 branch is updated from prime/develop?

It branches off here:

```
commit f85d29789eb6e3e95baec38de7c730812f84001d (prime/develop, origin/develop, origin/HEAD, develop)
Merge: 08c07d2 4c0780e
Author: Alan Garny <agarny@hellix.com>
Date:   Tue Oct 3 15:06:43 2017 +0200

    Remove XML format option (closes #209).
```

Which is a while back. I'm happy to rebase the branch to the latest prime/develop though

MichaelClerx commented:
@hsorby according to this graph (https://github.com/cellml/libcellml/network) you added a commit (https://github.com/hsorby/libcellml/commit/fa4e78d1f20a2005d53cb7c779534b562860b4f8) to branch `issue208` which is an obsolete branch I want to delete. The commit should have been at the end of branch `issue161` which is the one used in this pull request.

Could you re-apply that final commit to this branch, so I can delete `issue208` and avoid further confusion?

agarny commented:
Just to be 100% certain, I have:
 - `git clone https://github.com/MichaelClerx/libcellml.git` + `git checkout issue161`, checked the log and I indeed have `042f5411c11be7dd144414f06982c62b054e3253`. From there, I was able to build just fine. When it comes to the tests, I am getting the same output as the one I posted above.
 - `git clone https://github.com/cellml/libcellml.git`. From there, I was able to both build and test things just fine.

From there, I have checked your copy of `tests/parser/parser.cpp` and it's different from that of the official `develop` branch, especially `TEST(Parser, invalidXMLElements)`. I have 'fixed' it locally and now all the tests pass in your `issue161` branch. So, although I haven't checked, it looks like your branch is out of sync with the official `develop` branch.

MichaelClerx commented:
Yes that's right! You can see in the network branch that it's a few months behind because this pull request has been open for so long :-)
One way of fixing it is to rebase this branch (disconnect it from develop a few months back and attach it to develop's most recent commit). I'd like to do that but only after I hear from @hsorby about the commit he did on branch208

nickerso commented:
The way the development workflow usually works is that you would keep your topic branch up-to-date with `prime/develop` as things evolve. Indeed, the first thing the CI system does when testing a PR is to check if the PR can be merged into `prime/develop` and it is the merged code that gets tested, not the topic branch directly. At least I believe that is how it is supposed to work. 

I'm a git entry level user, so I never go near rebase :)

MichaelClerx commented:
That's quite cool! 

~I'll try and cherry pick that commit Hugh did and then rebase, so that this branch works when you check it out again...~

I've copy-pasted the changes Hugh made in his fork, and just continue like that!

Then maybe I can do some googling on windows and cmake and python debug... oh boy!

MichaelClerx commented:
Rebased now!
Remaining changes between this branch and prime/develop are:

1. Addition of all the python bindings
2. Changes to various CMakeLists.txt involving libXML. Not sure why they're on this branch, the python bindings don't really involve libXML...

Interestingly the non-python tests all fail as well on windows (they just say `***Failed`, no further error message)

MichaelClerx commented:
Ok, have separated the two issues into two separate branches now. The windows tests still don't pass, but now it's only the python tests that fail (as opposed to all tests)

MichaelClerx commented:
A major difference between the output logs from the windows and ubuntu/osx machines is that the windows machine runs Python3.4 (although 3.6 is also mentioned) while the ubuntu and os/x tests are run with Python2.7.

On my own machine the tests fail for Python3 too (Update: probably because I compiled it for python2, need to find out if both are supported by the current code!)

MichaelClerx commented:
Ok, requiring Python 3.4 or greater means the ubuntu and mac machines no longer build, but doesn't change anything for the windows case. So reverting that change!

MichaelClerx commented:
One of the first errors (when running ctest) is:

```
12: Test command: C:\Python34\python.exe "test_component.py"
12: Environment variables: 
12:  PATH=D:/buildslave_libcellml/Unit_Tests_Builder_Windows_8_64_Bit/libcellml-build/src/Debug;D:/buildslave_libcellml/Unit_Tests_Builder_Windows_8_64_Bit/libcellml-build/src/bindings/python/Debug;%PATH%
12:  PYTHONPATH=D:/buildslave_libcellml/Unit_Tests_Builder_Windows_8_64_Bit/libcellml-build/src/bindings/python/Debug
12: Test timeout computed to be: 9.99988e+006
12: E
12: ======================================================================
12: ERROR: test_component (__main__.ComponentTestCase)
12: ----------------------------------------------------------------------
12: Traceback (most recent call last):
12:   File "D:\buildslave_libcellml\Unit_Tests_Builder_Windows_8_64_Bit\libcellml-build\src\bindings\python\Debug\libcellml\component.py", line 14, in swig_import_helper
12:     return importlib.import_module(mname)
12:   File "C:\Python34\lib\importlib\__init__.py", line 109, in import_module
12:     return _bootstrap._gcd_import(name[level:], package, level)
12:   File "<frozen importlib._bootstrap>", line 2254, in _gcd_import
12:   File "<frozen importlib._bootstrap>", line 2237, in _find_and_load
12:   File "<frozen importlib._bootstrap>", line 2224, in _find_and_load_unlocked
12: ImportError: No module named 'libcellml._component'
```

The relevant part from the compilation logs is:

```
Project "D:\buildslave_libcellml\Unit_Tests_Builder_Windows_8_64_Bit\libcellml-build\ALL_BUILD.vcxproj" (1) is building "D:\buildslave_libcellml\Unit_Tests_Builder_Windows_8_64_Bit\libcellml-build\src\bindings\python\_component.vcxproj" (5) on node 1 (default targets).
PrepareForBuild:
  Creating directory "_component.dir\Debug\".
  Creating directory "_component.dir\Debug\_component.tlog\".
InitializeBuildStatus:
  Creating "_component.dir\Debug\_component.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.
CustomBuild:
  Building Custom Rule D:/buildslave_libcellml/Unit_Tests_Builder_Windows_8_64_Bit/libcellml/src/bindings/python/CMakeLists.txt
  CMake does not need to re-run because D:\buildslave_libcellml\Unit_Tests_Builder_Windows_8_64_Bit\libcellml-build\src\bindings\python\CMakeFiles\generate.stamp is up-to-date.
  Swig source
ClCompile:
  C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\x86_amd64\CL.exe /c /IC:\Python34\include /ID:\buildslave_libcellml\Unit_Tests_Builder_Windows_8_64_Bit\libcellml\src\api /I"D:\buildslave_libcellml\Unit_Tests_Builder_Windows_8_64_Bit\libcellml-build\src\api" /ID:\buildslave_libcellml\Unit_Tests_Builder_Windows_8_64_Bit\libcellml\src\api\libcellml\module /ID:\buildslave_libcellml\libraries\libxml2\include\libxml2 /Zi /nologo /W3 /WX- /Od /Ob0 /D WIN32 /D _WINDOWS /D _DEBUG /D SWIG_PYTHON_INTERPRETER_NO_DEBUG /D "CMAKE_INTDIR=\"Debug\"" /D _component_EXPORTS /D _WINDLL /D _MBCS /Gm- /EHsc /RTC1 /MDd /GS /fp:precise /Zc:wchar_t /Zc:forScope /Zc:inline /GR /Fo"_component.dir\Debug\\" /Fd"_component.dir\Debug\vc140.pdb" /Gd /TP /errorReport:queue "D:\buildslave_libcellml\Unit_Tests_Builder_Windows_8_64_Bit\libcellml-build\src\bindings\python\Debug\libcellml\componentPYTHON_wrap.cxx"
  componentPYTHON_wrap.cxx
Link:
  C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\x86_amd64\link.exe /ERRORREPORT:QUEUE /OUT:"D:\buildslave_libcellml\Unit_Tests_Builder_Windows_8_64_Bit\libcellml-build\src\bindings\python\Debug\_component.pyd" /INCREMENTAL /NOLOGO kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib C:\Python34\libs\python34.lib ..\..\Debug\cellmld.lib D:\buildslave_libcellml\libraries\libxml2\lib\xml2.lib /MANIFEST /MANIFESTUAC:"level='asInvoker' uiAccess='false'" /manifest:embed /DEBUG /PDB:"D:/buildslave_libcellml/Unit_Tests_Builder_Windows_8_64_Bit/libcellml-build/src/bindings/python/Debug/_component.pdb" /SUBSYSTEM:CONSOLE /TLBID:1 /DYNAMICBASE /NXCOMPAT /IMPLIB:"D:/buildslave_libcellml/Unit_Tests_Builder_Windows_8_64_Bit/libcellml-build/src/bindings/python/Debug/_component.lib" /MACHINE:X64  /machine:x64 /DLL _component.dir\Debug\componentPYTHON_wrap.obj
     Creating library D:/buildslave_libcellml/Unit_Tests_Builder_Windows_8_64_Bit/libcellml-build/src/bindings/python/Debug/_component.lib and object D:/buildslave_libcellml/Unit_Tests_Builder_Windows_8_64_Bit/libcellml-build/src/bindings/python/Debug/_component.exp
  _component.vcxproj -> D:\buildslave_libcellml\Unit_Tests_Builder_Windows_8_64_Bit\libcellml-build\src\bindings\python\Debug\_component.pyd
FinalizeBuildStatus:
  Deleting file "_component.dir\Debug\_component.tlog\unsuccessfulbuild".
  Touching "_component.dir\Debug\_component.tlog\_component.lastbuildstate".
Done Building Project "D:\buildslave_libcellml\Unit_Tests_Builder_Windows_8_64_Bit\libcellml-build\src\bindings\python\_component.vcxproj" (default targets).
```

That looks like it's built a file
```
D:\buildslave_libcellml\Unit_Tests_Builder_Windows_8_64_Bit\libcellml-build\src\bindings\python\Debug\_component.pyd
```
which it later can't find in 
```
PYTHONPATH=D:/buildslave_libcellml/Unit_Tests_Builder_Windows_8_64_Bit/libcellml-build/src/bindings/python/Debug
```

MichaelClerx commented:
This might be relevant: https://stackoverflow.com/questions/8846480/error-importing-a-pyd-file-as-a-python-module-from-a-pyo-file

MichaelClerx committed:
Removed remnants of windows/libxml changes directly on bindings branch

MichaelClerx commented:
All tests pass!

Would love some reviews so that we can merge, @agarny @nickerso @hsorby 

Please note: I had to rebase on top of the libxml changes, so if you've already checked out a version of this branch please delete it and check it out anew!

nickerso commented:
I've managed to get the latest changes building and the tests pass for me on Mac OS 10.13.3 using Python 2.7.10 (I think).

MichaelClerx commented:
It's currently set to build for python2! You can have it build for Python3 instead, but then our test machines fail. Will have to figure out how to distribute two versions via `pip` eventually, but that's a new ticket and a new PR :-)

nickerso reviewed (approved):
I'm not enough of a python expert to critically evaluate the bindings implementation, but the test code looks good and all the tests pass (python and C++) so I'm happy for this to be merged in. We can always reevaluate the implementation if needed after we see how the bindings work in the wild.

agarny commented:
@hsorby and @metatoaster, I wanted to add you as reviewers, but you are not current/former CellML editors, so I can't. Still, it would be nice to have your input from a pure Python viewpoint.

metatoaster commented on src/bindings/python/tests/test_imported_entity.py:
```diff
@@ -19,13 +19,15 @@ def test_named_entity(self):
         x = ImportedEntity()
         idx = 'test'
-        self.assertEqual(x.getId(), '')
+        self.assertIs(x.getId(), '')
```
Should really keep the use of `assertEqual` for any non-identity tests (i.e. test for `None`, or that the value assigned to one name is also to another name).

MichaelClerx replied:
Could you explain a bit more? I don't understand what you're saying here!

metatoaster replied:
`assertIs` is used to assert the exact same object (e.g. two assignments (variables) that point to the same object), and `assertEqual` asserts actual equality which is what is usually wanted.  See [this](https://stackoverflow.com/questions/7281774/whats-the-difference-between-assertequal-and-assertis-assertis-was-introduced) and [this](https://stackoverflow.com/questions/1504717/why-does-comparing-strings-in-python-using-either-or-is-sometimes-produce?noredirect=1&lq=1) stackoverflow posts.

MichaelClerx replied:
Thanks! I see what you mean now.

I think I was a bit worried that a `False` or a `[]` would get through, as in `if not '': print('hi')` and `if not False: print('hi')`, but on closer inspection those semantics don't apply here.

Will change!

MichaelClerx replied:
Done!

metatoaster commented on tests/bindings/python/test_basics.py:
```diff
@@ -1,11 +1,69 @@
+        try:
+            import libcellml
+        except ImportError as e:
+            self.fail(repr(e))
```
consider using [`assertRaises`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertRaises) to simplify readability. e.g.

```python
    def test_basics(self):
        with self.assertRaises(ImportError): 
            import libcellml
```

Likewise for below.  Though given the nature of the following tests, I would have a list of modules and iterate through them and import using `importlib` instead.  Though this explicit/wordy nature isn't bad within testing context.

MichaelClerx replied:
Cheers! Didn't know the asserts could be used as context managers. Will change!

MichaelClerx replied:
Hold on, the point here was just to assert it could be imported at all. Not that it raises a certain error. Will remove this whole test as others cover it too!

MichaelClerx replied:
Done!

metatoaster commented on tests/bindings/python/test_model.py:
```diff
@@ -38,8 +38,163 @@ def test_model(self):
+        del(m, u)
+        
+        # bool removeUnits(size_t index)
+        m = Model()
```
Why not break it down into different `test_model_feature` so that each test can be better documented and we don't need to worry about deleting variables.  This also allow tests to be independent of each other so that specific failure cases can be more easily identified when they happen.

MichaelClerx replied:
Ok!

MichaelClerx replied:
Done!

metatoaster commented on tests/bindings/python/test_units.py:
```diff
@@ -0,0 +1,71 @@
+        z = Units(y)
+        del(y,z)
+        
+        # Test inheritance
```
Ditto here, I would create a new function, e.g. `test_model_inheritance` or the like.

metatoaster commented on tests/bindings/python/test_units.py:
```diff
@@ -0,0 +1,71 @@
+        from libcellml.entity import Entity
+        self.assertIsInstance(x, Entity)
+        
+        # Test access to inherited methods
```
Again, `test_method_inheritance` or something to that effect, and so on so forth.

metatoaster commented:
`PYTHONPATH` is not an acceptable way to have Python be able to make use of these bindings.  Please generate a working `setup.py` such that the creation of Python wheels (or binary eggs) is possible for distribution.  Ultimately this should be as easy to install as [`python-libsbml`](https://pypi.python.org/pypi/python-libsbml/5.16.0) - note the many binary distribution options.  I am not saying to get those wheels up on PyPI now, but having a way to generate the wheel(s) first will be important.

The `_module.so` files should be generated into `libcellml` directory, to keep everything self-contained and not necessarily pollute the `site-packages` directory when this ends up being a proper Python package.

Anyway, as for the method to achieve this, do note that `libsbml` is a self-contained package that embeds the full bindings code and the Python stuff wraps around the whole build chain, so that is not going to help here.  This [example project](https://github.com/danielunderwood/swig-example) is more aligned with what this project has, however their generated `setup.py` directly installs the .so into the Python library location; this works, but it isn't desirable for the case in an artifact (wheel) build process when all resulting files are to be packaged into a usable form inside a wheel.

MichaelClerx commented:
Thanks Tommy!

> PYTHONPATH is not an acceptable way to have Python be able to make use of these bindings.

Agreed 100%!

But we do have a separate issue for this: #206
I'd very much like to merge these bindings before that is implemented, as it's a major change and I don't want this branch to go out of sync with the main develop branch again

metatoaster commented:
Ah, right. In any case, the Python bindings do seem to work as is. As far as I can see, the patches address the scope of the issue of getting the base `libcellml` bindings working.

MichaelClerx committed:
Changed assertIs to assertEqual

agarny commented:
Hmm, the Windows build failed because of `git` having not been recognised...?! @hsorby, would you know anything about this?

agarny commented:
FWIW, here is a direct link to the Windows build log: http://autotest.bioeng.auckland.ac.nz/libcellml-buildbot/builders/Unit%20Tests%20Builder%20Windows%208%2064%20Bit/builds/580/steps/Git%20Other/logs/stdio.

hsorby commented:
Yes, I have attempted a fix and @nickerso will test it out shortly?

nickerso commented:
All tests are passing and looks like most of the comments from @metatoaster have been addressed, apart from those bumped to other issues. I think we are good to merge this in now and refine things in future if needed.

agarny commented:
Agreed. I am just going to go through it this afternoon and make sure that I can get everything to build on my Windows, Linux and macOS machines. From there, I will merge it in.

agarny committed:
Prevent some warnings when building the documentation.

Indeed, we used to get three warnings that read `Duplicate explicit target name: "here"`.
```

agarny commented:
FWIW, the documentation should be updated to include some information about SWIG. I had it installed on my Linux and macOS machines, but not on my Windows one. Still, it can be retrieved from [here](http://www.swig.org/download.html).

This aside, I am happy, in the sense that it's building (both code and documentation) and testing fine on Windows, Linux and macOS. So, will merge.

MichaelClerx commented:
Hi all! I still need to address a few of @metatoaster 's comments. Should I do that in a new branch then?

nickerso commented:
yeah, would be good to make specific issues for the outstanding issues (grouped if that makes sense?) and then address them in separate pull requests.
